### PR TITLE
validation.py: Annotate function args and return values

### PR DESCRIPTION
**Pull request by elisno**
Starting to fix strict mypy errors as discussed in #307.

This PR handles the internal.validation module:

I ran:
```bash
mypy --install-types --non-interactive --strict --allow-untyped-defs --allow-untyped-calls --allow-any-generics --show-error-codes --exclude cleanlab/experimental cleanlab/internal/validation.py
```

### Errors before PR

```
cleanlab/internal/validation.py:61: error: Argument 1 to "len" has incompatible type "Union[ndarray[Any, Any], generic]"; expected "Sized"  [arg-type]
cleanlab/internal/validation.py:72: error: Tuple index out of range  [misc]
cleanlab/internal/validation.py:74: error: Tuple index out of range  [misc]
Found 3 errors in 1 file (checked 1 source file)
```

### After PR

```
Success: no issues found in 1 source file
```

## Some thoughts

- I think (X, y) might need some custom Union type to handle both numpy arrays and pandas dataframe, etc.
- All of the "assert" functions return None.

### CODE REVIEW

**Comment by jwmueller** [comment]
For args that can be multiple types, I believe you can do: `from typing import Union`
and then declare `Union[Type1, Type2]` as their type, or even `Optional[Union[Type1, Type2]`.

File: cleanlab/internal/validation.py
Comment (jwmueller) [inline cleanlab/internal/validation.py:28]: Let's make this type `Any` because we want to support as many data formats as possible.

File: cleanlab/internal/validation.py
Comment (jwmueller) [inline cleanlab/internal/validation.py:29]: the possible valid types for `y` would include: 
list, np.ndarray, np.generic, pd.Series, pd.DataFrame

File: cleanlab/internal/validation.py
Comment (jwmueller) [inline cleanlab/internal/validation.py:110]: Type of `X` should by `Any`

File: cleanlab/internal/validation.py
Comment (jwmueller) [inline cleanlab/internal/validation.py:134]: the possible valid types for `y` would include:
list, np.ndarray, np.generic, pd.Series, pd.DataFrame

File: cleanlab/internal/validation.py
Comment (jwmueller) [inline cleanlab/internal/validation.py:104]: Type of `X` should by `Any`

File: cleanlab/internal/validation.py
Comment (elisno) [reply cleanlab/internal/validation.py:29]: mypy isn't happy about using `len(y: np.generic)`, as it's undefined.

Comment (elisno) [reply cleanlab/internal/validation.py]: I'll start with everything except np.generic for now.

I'm trying not to change anything in the function bodies to avoid breaking any assumptions that are made in the rest of the codebase.

File: cleanlab/internal/validation.py
Comment (anishathalye) [inline cleanlab/internal/validation.py:28]: I understand why we're using `Any` here, because we don't want to spell out all the different possibilities / we want to allow certain duck-typed things without having Cleanlab depend on extra libraries, but "Any" doesn't read great in type annotations. What if we create a type alias for this, perhaps in a new `cleanlab.typing` module, and call it `dataset_like` (which will just be Any):

```python
from typing import TypeAlias

dataset_like: TypeAlias = Any
```

File: cleanlab/internal/validation.py
Comment (anishathalye) [inline cleanlab/internal/validation.py:29]: Maybe worth introducing an alias for this too?

```python
label_like: TypeAlias = Union[list, np.ndarray, pd.Series, pd.DataFrame]
```

File: cleanlab/internal/validation.py
Comment (anishathalye) [inline cleanlab/internal/validation.py:110]: If we make the above change, this will also change to `dataset_like`

File: cleanlab/internal/validation.py
Comment (anishathalye) [inline cleanlab/internal/validation.py:134]: `label_like`

File: cleanlab/internal/validation.py
Comment (anishathalye) [inline cleanlab/internal/validation.py:139]: would want to update this as well

**Review by anishathalye — COMMENTED** [review]
Looks great overall; only one recommendation: let's consider using type aliases to increase clarity and avoid code duplication.

**Review by anishathalye — COMMENTED** [review]
Looking good! One final question/change requested.

@jwmueller, what do you think of this `LabelLike` / `DatasetLike` plan? Also, if we ever decide to include all the libraries like TensorFlow etc. (which we probably won't), we can always narrow the definition of `LabelLike` from `Any` to the more specific `Union[...]`.

**Comment by jwmueller** [comment]
@anishathalye @elisno I also vote for the `LabelLike` / `DatasetLike` plan, where all `Union` and `Any` statements are deferred to the definition of the  `LabelLike` / `DatasetLike` types, so that the function args solely say:

``` X : DatasetLike, 
      y : LabelLike,
```

**Comment by jwmueller** [comment]
@anishathalye should the type names actually be lower-case like in your earlier comments?  I.e. 
```
dataset_like, label_like
```
vs.
```
DatasetLike, LabelLike
```

I'm not sure which is more Pythonic, but I've definitely encountered the lower-case ones more frequently.

**Comment by elisno** [comment]
Glad you pointed out the capitalization. I used this as a reference


https://peps.python.org/pep-0484/#type-aliases

Maybe we want to look into `TypeVar` for parametrizing the generic types?

File: cleanlab/typing.py
Comment (anishathalye) [inline cleanlab/typing.py:5]: Can we type these as type aliases? (https://github.com/cleanlab/cleanlab/pull/317#discussion_r927905542)

File: cleanlab/typing.py
Comment (anishathalye) [inline cleanlab/typing.py:9]: Same thing as above here.

**Review by anishathalye — COMMENTED** [review]
@jwmueller I think Python isn't super consistent with formatting here. Many of the types come from class names, so those are upper camel case, and the typing package seems to use upper camel case as well, so I'd prefer that here (so `LabelLike` and `DatasetLike`).

As far as generics, maybe that will be nice to add at some point, but we could do that in a separate PR / is lower priority than getting typing coverage over the whole package (have it pass `mypy --strict`).

File: cleanlab/typing.py
Comment (elisno) [reply cleanlab/typing.py:5]: Only for Python 3.10.

Comment (elisno) [reply cleanlab/typing.py:5]: https://docs.python.org/3/library/typing.html#typing.TypeAlias

Comment (elisno) [reply cleanlab/typing.py:5]:
```python
>>> import sys
>>> sys.version
'3.8.5 (default, Sep  4 2020, 07:30:14) \n[GCC 7.3.0]'
>>> from typing import TypeAlias
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'TypeAlias' from 'typing'
```

Comment (elisno) [reply cleanlab/typing.py:5]: @anishathalye Did you have something like this in mind?

```python
# Import TypeAlias on Python 3.10+
try:
    from typing import TypeAlias
except ImportError:
    TypeAlias = type(None)


LabelLike: TypeAlias = Union[list, np.ndarray, pd.Series, pd.DataFrame]
DatasetLike: TypeAlias = Any
```

Comment (anishathalye) [reply cleanlab/typing.py:5]: Ah I didn't realize that was only added in 3.10. We want to support 3.6+ without too much compatibility code, so let's stick with the current code in the PR.

File: cleanlab/internal/validation.py
Comment (anishathalye) [reply cleanlab/internal/validation.py:149]: Looking at the docs for `pd.Series.values`, [pd.Series.to_numpy()](https://pandas.pydata.org/docs/reference/api/pandas.Series.to_numpy.html#pandas.Series.to_numpy) might be better to use here.

File: cleanlab/internal/validation.py
Comment (anishathalye) [reply cleanlab/internal/validation.py]: No, I don't think we need type aliases for those. Probabilities are floats (that are between 0 and 1), scores are floats (that are between 0 and 1), a joint is a 2d numpy array of floats.

File: cleanlab/internal/validation.py
Comment (elisno) [reply cleanlab/internal/validation.py:149]:
```python
return y.values.to_numpy()
```
these mypy errors

```
cleanlab/internal/validation.py:151: error: Returning Any from function declared to return "ndarray[Any, Any]"
cleanlab/internal/validation.py:151: error: Item "ndarray[Any, Any]" of "Union[ExtensionArray, ndarray[Any, Any]]" has no attribute "to_numpy"
```

Comment (anishathalye) [reply cleanlab/internal/validation.py:149]: Oh I think it's just `y.to_numpy()`.

Comment (elisno) [reply cleanlab/internal/validation.py:149]: Down to one error:

```
cleanlab/internal/validation.py:151: error: Returning Any from function declared to return "ndarray[Any, Any]"
```

Comment (elisno) [reply cleanlab/internal/validation.py:149]: Maybe turn this return value into a variable and annotate it explicitly?

Comment (elisno) [reply cleanlab/internal/validation.py:149]: Similary to what is done here:
https://github.com/jwmueller/cleanlab/blob/a0e1227ae5657e520e106552360072c39ea7fccb/cleanlab/rank.py#L647

Comment (elisno) [reply cleanlab/internal/validation.py:149]: This works:

```python
if isinstance(y, pd.Series):
        y_series: np.ndarray = y.to_numpy()
        return y_series
```

**Review by anishathalye — APPROVED** [review]
LGTM! Thanks for the PR. Will wait for Jonas to click the merge button.

**Review by jwmueller — APPROVED** [review]
LGTM, thanks for the useful contribution and your many good suggestions!

**Pull request merged**
